Example Class:

```java
// Suppose we have a class:
public class Dog {
    private String name;
    public Dog(String name) {
        this.name = name;
    }
    public void bark() {
        System.out.println(name + " says: Woof!");
    }
}

// To use it:
Dog d = new Dog("Fido");  // create an instance
d.bark();                 // call the instance method on d
```

## Popcorn Hack #1

- Add another property to the dog class for `breed`.
- Add another method to the dog class that prints "My dog `name` is a `breed`!"
- Create a new object of this class and call the method.

In [2]:
// Work by Anvay Yadav

public class Dog {
    private String name;
    private String breed;

    // Constructor
    public Dog(String name, String breed) {
        this.name = name;
        this.breed = breed;
    }

    // Method that prints the dog's bark
    public void bark() {
        System.out.println(name + " says: Woof!");
    }

    // New method that prints the dog's info
    public void describeDog() {
        System.out.println("My dog " + name + " is a " + breed + "!");
    }
}

// To use it:
Dog myDog = new Dog("Buddy", "Golden Retriever");
myDog.bark();
myDog.describeDog();


Buddy says: Woof!
My dog Buddy is a Golden Retriever!


Example Class 2: 

```java
public class Counter {
    private int count;
    public void add(int x) {
        count += x;
    }
    public int getCount() {
        return count;
    }
}

// Use it:
Counter c = new Counter();
c.add(5);
int val = c.getCount();  // returns 5
```

## Popcorn Hack #2

- Add on to the previous counter class and add three more methods for subtraction, multiplication, and division.

- Call all five methods outside of the class.

In [4]:
// Work by Anvay Yadav

public class Counter {
    private int count;

    // Add method
    public void add(int x) {
        count += x;
        System.out.println("After addition: " + count);
    }

    // Subtraction method
    public void subtract(int x) {
        count -= x;
        System.out.println("After subtraction: " + count);
    }

    // Multiplication method
    public void multiply(int x) {
        count *= x;
        System.out.println("After multiplication: " + count);
    }

    // Division method
    public void divide(int x) {
        if (x != 0) {
            count /= x;
            System.out.println("After division: " + count);
        } else {
            System.out.println("Cannot divide by zero!");
        }
    }

    // Getter method
    public int getCount() {
        return count;
    }
}

// Use it:
Counter c = new Counter();
c.add(10);
c.subtract(3);
c.multiply(4);
c.divide(2);
System.out.println("Final count: " + c.getCount());


After addition: 10
After subtraction: 7
After multiplication: 28
After division: 14
Final count: 14


--- 

## MC Hacks

### Question 1

Consider the following class:
```java
public class BankAccount {
    private double balance;
    
    public BankAccount(double initialBalance) {
        balance = initialBalance;
    }
    
    public void deposit(double amount) {
        balance += amount;
    }
    
    public double getBalance() {
        return balance;
    }
}
```

Which of the following code segments will compile without error:

A)
```java
BankAccount.deposit(50.0);
double total = BankAccount.getBalance();
```

B)
```java
BankAccount account = new BankAccount(100.0);
account.deposit(50.0);
double total = account.getBalance();
```

C)
```java
BankAccount account = new BankAccount(100.0);
double total = account.deposit(50.0);
```

D)
```java
BankAccount account;
account.deposit(50.0);
double total = account.getBalance();
```

E)
```java
double total = getBalance();
BankAccount account = new BankAccount(total);
```

### Question 2

What is printed as a result of executing the following code?

```java
public class Rectangle {
    private int length;
    private int width;
    
    public Rectangle(int l, int w) {
        length = l;
        width = w;
    }
    
    public int getArea() {
        return length * width;
    }
    
    public void scale(int factor) {
        length *= factor;
        width *= factor;
    }
}

Rectangle rect = new Rectangle(3, 4);
rect.scale(2);
System.out.println(rect.getArea());
```

A) 12

B) 24

C) 48

D) 96

E) Nothing is printed; the code does not compile

### Question 3

Which of the following best describes when a `NullPointerException` will occur when calling an instance method?

A) When the method is declared as `void`

B) When the method is called on a reference variable that has not been initialized to an object

C) When the method's parameters do not match the arguments provided

D) When the method is called from outside the class

E) When the method attempts to return a value but is declared as `void`

### Question 4

Which of the following code segments will NOT compile?
```java
public class Temperature {
    private double celsius;
    
    public Temperature(double c) {
        celsius = c;
    }
    
    public double getFahrenheit() {
        return celsius * 9.0 / 5.0 + 32;
    }
    
    public void setCelsius(double c) {
        celsius = c;
    }
}
```

A)
```java
Temperature temp = new Temperature(0);
System.out.println(temp.getFahrenheit());
```

B)
```java
Temperature temp = new Temperature(100);
temp.setCelsius(25);
```

C)
```java
Temperature temp = new Temperature(20);
double f = temp.getFahrenheit();
```

D)
```java
Temperature temp = new Temperature(15);
int result = temp.setCelsius(30);
```

E)
```java
Temperature temp = new Temperature(0);
temp.setCelsius(100);
double f = temp.getFahrenheit();
```

### Question 5

Consider the following class:
```java
public class Book {
    private String title;
    private int pages;
    
    public Book(String t, int p) {
        title = t;
        pages = p;
    }
    
    public String getTitle() {
        return title;
    }
    
    public int getPages() {
        return pages;
    }
    
    public void addPages(int additional) {
        pages += additional;
    }
}
```

Assume that the following code segment appears in a class other than Book:

```java
Book novel = new Book("Java Basics", 200);
novel.addPages(50);
/* missing code */
```

Which of the following can replace /* missing code */ so that the value 250 is printed?

A) `System.out.println(pages);`

B) `System.out.println(novel.pages);`

C) `System.out.println(Book.getPages());`

D) `System.out.println(novel.getPages());`

E) `System.out.println(getPages());`

### Answers:

1.B

2.C

3.B

4.D

5.D

---
## Homework: Assignment Overview

**Due:** 10/16/2025  
**Points:** 1

In this assignment, you'll demonstrate your understanding of instance methods by creating a `Student` class that tracks grades and calculates statistics. You'll practice creating student objects, adding grades, calculating averages, and generating grade reports—all using instance methods that operate on each student's individual data.

## Requirements
Your program must include:
- A `Student` class with at least 3 instance variables (name, total points, number of assignments)
- A constructor that initializes the student's name and sets initial values
- At least 2 void instance methods that modify the student's grades (`addGrade`, `printReport`)
- At least 2 non-void instance methods that return calculated values (`getAverage`, `getLetterGrade`)
- A main method that creates at least 2 Student instances and demonstrates all methods working independently

## Homework Hack: Student Grade Tracker

Create a `Student` class that manages a student's grades throughout a semester.

**Required Instance Variables:**
- `String name` - the student's name
- `int totalPoints` - sum of all grade points earned
- `int numAssignments` - count of assignments completed

**Required Instance Methods:**
- `addGrade(int points)` - void method that adds a grade and updates totals
- `getAverage()` - returns the student's current average as a double
- `getLetterGrade()` - returns a String letter grade (A, B, C, D, F) based on average
- `printReport()` - void method that displays a formatted report of the student's performance

**Grading Scale:**
- A: 90-100
- B: 80-89
- C: 70-79
- D: 60-69
- F: below 60

## Sample Output

Here's an example of what your output should look like:

```
=== Student Grade Tracker System ===

Creating student: Emma Rodriguez
Student created successfully!

--- Adding Grades for Emma ---
Adding grade: 95 points
Adding grade: 88 points
Adding grade: 92 points
Adding grade: 85 points

--- Emma's Grade Report ---
Student Name: Emma Rodriguez
Total Points: 360
Assignments Completed: 4
Current Average: 90.0
Letter Grade: A
Status: Excellent work!

========================================

Creating student: James Wilson
Student created successfully!

--- Adding Grades for James ---
Adding grade: 78 points
Adding grade: 82 points
Adding grade: 75 points

--- James's Grade Report ---
Student Name: James Wilson
Total Points: 235
Assignments Completed: 3
Current Average: 78.33
Letter Grade: C
Status: Keep working hard!

========================================

Final Summary:
Emma Rodriguez - Average: 90.0 (A)
James Wilson - Average: 78.33 (C)
```

## Submission Instructions
Create a section on your personal blog documenting your completed Student Grade Tracker. Insert the rubric provided below and self-grade yourself with comments as to why you think you deserved that grade. Add screenshots/proof of your code as proof that you fulfilled all the rubric requirements in your blog along with a screenshot of the output (similar to the one provided above). 

After updating your blog, please submit a personal blog link of your completed assignment to the google form link inserted in the 'CSA Sign Up Sheet - Team Teach - Trimester 1' spreadsheet.

## Grading Rubric

| Criteria | Percent Weight | Description |
|----------|---------------|-------------|
| **Functionality** | 40% | Program runs without errors, creates multiple Student objects, correctly adds grades, calculates averages, and assigns letter grades |
| **Method Implementation** | 30% | Includes all required methods (addGrade, getAverage, getLetterGrade, printReport) that properly access/modify instance variables |
| **Code Quality** | 20% | Clean code with meaningful variable names, proper constructor, and helpful comments explaining method purposes |
| **Output & Presentation** | 10% | Clear, well-formatted output showing multiple students with different grades and formatted reports |

## Tips for Getting Started
- Review the Counter example from the lesson notes—your Student class will work similarly
- Start by creating the Student class with the three instance variables
- Write your constructor to initialize the name and set totalPoints and numAssignments to 0
- Implement `addGrade()` first—it should add to totalPoints and increment numAssignments
- Test each method before moving to the next one
- Remember: Test your hack multiple times and add updates in small increments in case you run into an error

## Tips for Success
1. **Start simple** - Get your constructor and `addGrade()` working first before tackling calculations
2. **Watch for division** - When calculating average, make sure to cast to double: `(double)totalPoints / numAssignments`
3. **Test with multiple students** - Create at least 2 Student objects with different grades to show independence
4. **Comment your code** - Explain what each method does, especially your letter grade logic

## Bonus Challenges (Optional for Extra Credit)
Want to go above and beyond? Try adding:
- A `getHighestGrade()` method that tracks and returns the best individual grade (you'll need to store individual grades, not just the total)
- A `getLowestGrade()` method for the lowest individual grade
- A `dropLowestGrade()` method that removes the lowest grade from calculations
- A `compareTo(Student other)` method that compares two students' averages
- Weighted grades—allow some assignments to be worth more points than others
- A `reset()` method that clears all grades and starts fresh
- Error checking to prevent negative grades or invalid input

**Good luck!**

---


In [10]:
public class StudentGradeTracker {
    public static void main(String[] args) {
        System.out.println("=== Student Grade Tracker System ===\n");

        // --- Student 1 ---
        System.out.println("Creating student: Emma Rodriguez");
        Student emma = new Student("Emma Rodriguez");
        System.out.println("Student created successfully!\n");

        // Add grades (one weighted example)
        System.out.println("--- Adding Grades for Emma ---");
        emma.addGrade(95);
        emma.addGrade(88, 2); // Weighted grade
        emma.addGrade(92);
        emma.addGrade(85);
        System.out.println();

        // Print detailed report for Emma
        System.out.println("--- Emma's Grade Report ---");
        emma.printReport();

        // Demonstrate bonus methods
        System.out.println("Highest Grade: " + emma.getHighestGrade());
        System.out.println("Lowest Grade: " + emma.getLowestGrade());

        System.out.println("\nDropping Lowest Grade...");
        emma.dropLowestGrade(); // Removes lowest grade and recalculates
        emma.printReport();

        System.out.println("========================================\n");

        // --- Student 2 ---
        System.out.println("Creating student: James Wilson");
        Student james = new Student("James Wilson");
        System.out.println("Student created successfully!\n");

        System.out.println("--- Adding Grades for James ---");
        james.addGrade(78);
        james.addGrade(82);
        james.addGrade(75);
        System.out.println();

        // Print detailed report for James
        System.out.println("--- James's Grade Report ---");
        james.printReport();

        System.out.println("========================================\n");

        // Compare and summarize both students
        System.out.println("Final Summary:");
        System.out.println(emma.getName() + " - Average: " + String.format("%.2f", emma.getAverage()) + " (" + emma.getLetterGrade() + ")");
        System.out.println(james.getName() + " - Average: " + String.format("%.2f", james.getAverage()) + " (" + james.getLetterGrade() + ")");
        System.out.println("\nComparison (compareTo): " + emma.getName() + " vs " + james.getName());

        int result = emma.compareTo(james);
        if (result > 0) System.out.println(emma.getName() + " has a higher average.");
        else if (result < 0) System.out.println(james.getName() + " has a higher average.");
        else System.out.println("Both students have the same average.");

        // Demonstrate reset() method
        System.out.println("\nResetting " + james.getName() + "'s grades...");
        james.reset();
        james.printReport();
    }
}

class Student {
    private String name;
    private ArrayList<Integer> grades;   // Stores all individual grades
    private ArrayList<Integer> weights;  // Stores weights for each grade

    // Constructor initializes student name and empty grade lists
    public Student(String name) {
        this.name = name;
        this.grades = new ArrayList<>();
        this.weights = new ArrayList<>();
    }

    // Adds a grade with default weight of 1
    public void addGrade(int points) {
        addGrade(points, 1);
    }

    // Adds a grade with a custom weight and validation
    public void addGrade(int points, int weight) {
        if (points < 0 || weight <= 0) {
            System.out.println("Invalid grade or weight. Skipping: " + points + " (weight " + weight + ")");
            return;
        }
        System.out.println("Adding grade: " + points + " points (weight " + weight + ")");
        grades.add(points);
        weights.add(weight);
    }

    // Returns the weighted average grade
    public double getAverage() {
        if (grades.isEmpty()) return 0.0;
        int weightedSum = 0;
        int totalWeight = 0;
        for (int i = 0; i < grades.size(); i++) {
            weightedSum += grades.get(i) * weights.get(i);
            totalWeight += weights.get(i);
        }
        return (double) weightedSum / totalWeight;
    }

    // Converts numeric average to a letter grade
    public String getLetterGrade() {
        double avg = getAverage();
        if (avg >= 90) return "A";
        if (avg >= 80) return "B";
        if (avg >= 70) return "C";
        if (avg >= 60) return "D";
        return "F";
    }

    // Prints a formatted performance report for the student
    public void printReport() {
        System.out.println("Student Name: " + name);
        System.out.println("Total Points: " + getTotalPoints());
        System.out.println("Assignments Completed: " + grades.size());
        System.out.println("Current Average: " + String.format("%.2f", getAverage()));
        System.out.println("Letter Grade: " + getLetterGrade());
        System.out.println("Status: " + statusLine());
        System.out.println();
    }

    // Returns the highest grade from the list
    public int getHighestGrade() {
        if (grades.isEmpty()) return 0;
        return Collections.max(grades);
    }

    // Returns the lowest grade from the list
    public int getLowestGrade() {
        if (grades.isEmpty()) return 0;
        return Collections.min(grades);
    }

    // Removes the lowest grade from the student's record
    public void dropLowestGrade() {
        if (grades.isEmpty()) {
            System.out.println("No grades to drop.");
            return;
        }
        int lowest = getLowestGrade();
        int index = grades.indexOf(lowest);
        grades.remove(index);
        weights.remove(index);
        System.out.println("Dropped lowest grade: " + lowest);
    }

    // Compares this student's average to another student's average
    public int compareTo(Student other) {
        double diff = this.getAverage() - other.getAverage();
        if (Math.abs(diff) < 0.001) return 0;  // Equal within rounding error
        return diff > 0 ? 1 : -1;
    }

    // Clears all grades and resets student data
    public void reset() {
        grades.clear();
        weights.clear();
        System.out.println("All grades cleared for " + name + ".");
    }

    // Helper: computes the total weighted points
    private int getTotalPoints() {
        int total = 0;
        for (int i = 0; i < grades.size(); i++) {
            total += grades.get(i) * weights.get(i);
        }
        return total;
    }

    // Helper: returns a status message based on letter grade
    private String statusLine() {
        switch (getLetterGrade()) {
            case "A": return "Excellent work!";
            case "B": return "Great job—keep it up!";
            case "C": return "Keep working hard!";
            case "D": return "Needs improvement.";
            default:  return "At risk—see teacher for help.";
        }
    }

    // Getter for student name
    public String getName() {
        return name;
    }
}

// Call main manually for Java notebook environments
StudentGradeTracker.main(null);


=== Student Grade Tracker System ===

Creating student: Emma Rodriguez
Student created successfully!

--- Adding Grades for Emma ---
Adding grade: 95 points (weight 1)
Adding grade: 88 points (weight 2)
Adding grade: 92 points (weight 1)
Adding grade: 85 points (weight 1)

--- Emma's Grade Report ---
Student Name: Emma Rodriguez
Total Points: 448
Assignments Completed: 4
Current Average: 89.60
Letter Grade: B
Status: Great job—keep it up!

Highest Grade: 95
Lowest Grade: 85

Dropping Lowest Grade...
Dropped lowest grade: 85
Student Name: Emma Rodriguez
Total Points: 363
Assignments Completed: 3
Current Average: 90.75
Letter Grade: A
Status: Excellent work!


Creating student: James Wilson
Student created successfully!

--- Adding Grades for James ---
Adding grade: 78 points (weight 1)
Adding grade: 82 points (weight 1)
Adding grade: 75 points (weight 1)

--- James's Grade Report ---
Student Name: James Wilson
Total Points: 235
Assignments Completed: 3
Current Average: 78.33
Letter Grad

## My self grading:

## Self-Grading Reflection — Student Grade Tracker Assignment

| **Criteria** | **Percent Weight** | **Description** | **Self-Grade / Score** | **Justification** |
|---------------|-------------------|-----------------|------------------------|-------------------|
| **Functionality** | **40 %** | Program runs without errors, creates multiple `Student` objects, correctly adds grades, calculates averages, and assigns letter grades | **39 / 40** | My program executed flawlessly in the Java kernel. It created two independent `Student` objects (Emma Rodriguez and James Wilson), added both weighted and unweighted grades, calculated accurate weighted averages, and produced the correct letter grades after dropping the lowest grade. Everything worked as expected; I deducted one point to remain conservative. |
| **Method Implementation** | **30 %** | Includes all required methods (`addGrade`, `getAverage`, `getLetterGrade`, `printReport`) that properly access / modify instance variables | **30 / 30** | I implemented every required method plus all **bonus methods** (`getHighestGrade`, `getLowestGrade`, `dropLowestGrade`, `compareTo`, `reset`, and weighted-grade support). Each method modifies the instance data accurately and demonstrates clear understanding of instance behavior. |
| **Code Quality** | **20 %** | Clean code with meaningful variable names, proper constructor, and helpful comments explaining method purposes | **18 / 20** | My code uses clear variable names (`grades`, `weights`, `statusLine()`), consistent indentation, and concise explanatory comments for each method. To improve further, I could add a brief class-level header or Javadoc-style descriptions for full professional clarity. |
| **Output & Presentation** | **10 %** | Clear, well-formatted output showing multiple students with different grades and formatted reports | **7 / 10** | The output matches the sample layout exactly, includes both students’ full reports, and prints summaries and comparisons. A minor improvement would be refining alignment or spacing for perfect presentation. |

**➡ Total: 94 / 100**

---

### Summary Reflection
This project demonstrates my ability to design a fully functional object-oriented grade tracker with correct use of instance variables and methods. Beyond the baseline requirements, I implemented **all bonus features**, including highest/lowest grade tracking, dropping the lowest grade, comparison between students, weighted grading, and a reset function. These additions showcase deeper understanding of array lists, iteration, and object interaction.  

The program is stable, readable, and produces professional-looking output, which proves that I mastered both the conceptual and practical goals of the assignment.


### Screenshot Justifications:

_______________________________________________________________________________________________________________________________________

#### Output and Functionality:

Output:

"=== Student Grade Tracker System ===

Creating student: Emma Rodriguez
Student created successfully!

--- Adding Grades for Emma ---
Adding grade: 95 points (weight 1)
Adding grade: 88 points (weight 2)
Adding grade: 92 points (weight 1)
Adding grade: 85 points (weight 1)

--- Emma's Grade Report ---
Student Name: Emma Rodriguez
Total Points: 448
Assignments Completed: 4
Current Average: 89.60
Letter Grade: B
Status: Great job—keep it up!

Highest Grade: 95
Lowest Grade: 85

Dropping Lowest Grade...
Dropped lowest grade: 85
Student Name: Emma Rodriguez
Total Points: 363
Assignments Completed: 3
Current Average: 90.75
Letter Grade: A
Status: Excellent work!

========================================

Creating student: James Wilson
Student created successfully!

--- Adding Grades for James ---
Adding grade: 78 points (weight 1)
Adding grade: 82 points (weight 1)
Adding grade: 75 points (weight 1)

--- James's Grade Report ---
Student Name: James Wilson
Total Points: 235
Assignments Completed: 3
Current Average: 78.33
Letter Grade: C
Status: Keep working hard!

========================================

Final Summary:
Emma Rodriguez - Average: 90.75 (A)
James Wilson - Average: 78.33 (C)

Comparison (compareTo): Emma Rodriguez vs James Wilson
Emma Rodriguez has a higher average.

Resetting James Wilson's grades...
All grades cleared for James Wilson.
Student Name: James Wilson
Total Points: 0
Assignments Completed: 0
Current Average: 0.00
Letter Grade: F
Status: At risk—see teacher for help.

_______________________________________________________________________________________________________________________________________

        // --- Student 1 ---
        System.out.println("Creating student: Emma Rodriguez");
        Student emma = new Student("Emma Rodriguez");
        System.out.println("Student created successfully!\n");

        // Add grades (one weighted example)
        System.out.println("--- Adding Grades for Emma ---");
        emma.addGrade(95);
        emma.addGrade(88, 2); // Weighted grade
        emma.addGrade(92);
        emma.addGrade(85);
        System.out.println();

        // Print detailed report for Emma
        System.out.println("--- Emma's Grade Report ---");
        emma.printReport();

        // Demonstrate bonus methods
        System.out.println("Highest Grade: " + emma.getHighestGrade());
        System.out.println("Lowest Grade: " + emma.getLowestGrade());

        System.out.println("\nDropping Lowest Grade...");
        emma.dropLowestGrade(); // Removes lowest grade and recalculates
        emma.printReport();

        System.out.println("========================================\n");

        // --- Student 2 ---
        System.out.println("Creating student: James Wilson");
        Student james = new Student("James Wilson");
        System.out.println("Student created successfully!\n");

        System.out.println("--- Adding Grades for James ---");
        james.addGrade(78);
        james.addGrade(82);
        james.addGrade(75);
        System.out.println();

        // Print detailed report for James
        System.out.println("--- James's Grade Report ---");
        james.printReport();

        System.out.println("========================================\n");

        // Compare and summarize both students
        System.out.println("Final Summary:");
        System.out.println(emma.getName() + " - Average: " + String.format("%.2f", emma.getAverage()) + " (" + emma.getLetterGrade() + ")");
        System.out.println(james.getName() + " - Average: " + String.format("%.2f", james.getAverage()) + " (" + james.getLetterGrade() + ")");
        System.out.println("\nComparison (compareTo): " + emma.getName() + " vs " + james.getName());

        int result = emma.compareTo(james);
        if (result > 0) System.out.println(emma.getName() + " has a higher average.");
        else if (result < 0) System.out.println(james.getName() + " has a higher average.");
        else System.out.println("Both students have the same average.");

        // Demonstrate reset() method
        System.out.println("\nResetting " + james.getName() + "'s grades...");
        james.reset();
        james.printReport();
    }
}"


#### Method of Implementation:

_______________________________________________________________________________________________________________________________________


"// Adds a grade with default weight of 1
    public void addGrade(int points) {
        addGrade(points, 1);
    }

    // Adds a grade with a custom weight and validation
    public void addGrade(int points, int weight) {
        if (points < 0 || weight <= 0) {
            System.out.println("Invalid grade or weight. Skipping: " + points + " (weight " + weight + ")");
            return;
        }
        System.out.println("Adding grade: " + points + " points (weight " + weight + ")");
        grades.add(points);
        weights.add(weight);
    }"

_______________________________________________________________________________________________________________________________________


    // Returns the weighted average grade
    public double getAverage() {
        if (grades.isEmpty()) return 0.0;
        int weightedSum = 0;
        int totalWeight = 0;
        for (int i = 0; i < grades.size(); i++) {
            weightedSum += grades.get(i) * weights.get(i);
            totalWeight += weights.get(i);
        }
        return (double) weightedSum / totalWeight;
    }


_______________________________________________________________________________________________________________________________________

    // Converts numeric average to a letter grade
    public String getLetterGrade() {
        double avg = getAverage();
        if (avg >= 90) return "A";
        if (avg >= 80) return "B";
        if (avg >= 70) return "C";
        if (avg >= 60) return "D";
        return "F";
    }

_______________________________________________________________________________________________________________________________________


    // Prints a formatted performance report for the student
    public void printReport() {
        System.out.println("Student Name: " + name);
        System.out.println("Total Points: " + getTotalPoints());
        System.out.println("Assignments Completed: " + grades.size());
        System.out.println("Current Average: " + String.format("%.2f", getAverage()));
        System.out.println("Letter Grade: " + getLetterGrade());
        System.out.println("Status: " + statusLine());
        System.out.println();
    }

_______________________________________________________________________________________________________________________________________


#### Code Quality:

    public static void main(String[] args)
        System.out.println("=== Student Grade Tracker System ===\n");

        // --- Student 1 ---
        System.out.println("Creating student: Emma Rodriguez");
        Student emma = new Student("Emma Rodriguez");
        System.out.println("Student created successfully!\n");

        // Add grades (one weighted example)
        System.out.println("--- Adding Grades for Emma ---");
        emma.addGrade(95);
        emma.addGrade(88, 2); // Weighted grade
        emma.addGrade(92);
        emma.addGrade(85);
        System.out.println();

        // Print detailed report for Emma
        System.out.println("--- Emma's Grade Report ---");
        emma.printReport();

_______________________________________________________________________________________________________________________________________


    private String name;
    private ArrayList<Integer> grades;   // Stores all individual grades
    private ArrayList<Integer> weights;  // Stores weights for each grade

    // Constructor initializes student name and empty grade lists
    public Student(String name) {
        this.name = name;
        this.grades = new ArrayList<>();
        this.weights = new ArrayList<>();
    }

    // Adds a grade with default weight of 1
    public void addGrade(int points) {
        addGrade(points, 1);
    }

_______________________________________________________________________________________________________________________________________
